### PR TITLE
ProtectedBranches API changes for Required Review Enforcement

### PR DESCRIPTION
**Pull request by M-Zuber**
Edited by @ryangribble

Hope you don't mind @M-Zuber but I've taken over this PR 😀 

I realized that [we don't have much time](https://developer.github.com/changes/2017-06-16-loki-preview-ending-soon/) before we have to pass the new `required_pull_request_reviews` field in all branch protection calls, otherwise they will stop working when the API becomes official in less than a week

There were also many changes to this API during it's preview period, and it now has functionality to control the dismissal of pull request reviews

I've implemented the extra fields required as well as the new explicit GET/UPDATE/REMOVE "Review Enforcement" client methods.

- [x] Add new dismissal related fields to `BranchProtectionRequiredReviews` and `BranchProtectionRequiredReviewsUpdate`
- [x] Update integration tests for parent branch protection calls to use/validate the new fields
- [x] Implement new explicit `GetReviewRestrictions`, `UpdateReviewRestrictions` and `RemoveReviewRestrictions` client methods\
- [x] Unit tests
- [x] Integration tests
- [x] Rinse and repeat for the `ObservableRepositoryBranchesClient`
---

~~This is on hold until #1598 lands, as that has a decent sized impact on the code~~
~~- [ ] Unit Tests~~
~~- [ ] Integration Tests~~

~~- [ ] Test if @ryangribble is correct [here](https://github.com/octokit/octokit.net/issues/1512#issuecomment-266735950)~~
~~- if yes~~
~~- [ ] Update Contsructors~~
~~- [ ] Update Unit tests~~
~~- [ ] Update integration tests~~

### CODE REVIEW

**Comment by ryangribble** [comment]
Looking pretty good so far :+1:

The names are getting pretty long though!  I wonder if instead of `BranchProtectionRequiredPullRequestReviewsUpdate` we can get away with `BranchProtectionRequiredReviewsUpdate` instead.  What do you think?

In terms of the "test whether @ryangribble is right" thing, Im pretty positive it's going to be how I said. 
eg if you look at the webUI below, passing the item as null controls the "require reviews before merging" checkbox, whilst passing the `IncludeAdmins` as true/false controls the Include Administrators checkbox.  If you always pass up the object, how will you ever configure branch protection to NOT require a review?

image

Tests will reveal soon enough I guess 😛

**Comment by M-Zuber** [comment]
> The names are getting pretty long though! I wonder if instead of `BranchProtectionRequiredPullRequestReviewsUpdate` we can get away with `BranchProtectionRequiredReviewsUpdate` instead. What do you think?

`BranchProtectionRequiredReviewsUpdate` sounds good enough to me

I am also pretty sure that you are correct, I just made my initial code based off of my understanding of the docs. I was a bit too daft to go check the UI 😆

**Comment by M-Zuber** [comment]
@ryangribble while I am fixing the naming for the new object, should I shorten the other names, or make that a separate pr?

**Comment by mkchandler** [comment]
@M-Zuber Any updates on this?

**Comment by M-Zuber** [comment]
Thank you for the nudge

Had a rough spot in my personal life, and then my computer needed to be repaved. 
I will do my best to finish this up by the end of the week. 
A decent amount of the code is already written, I just have to polish and push

**Comment by mkchandler** [comment]
@M-Zuber No worries! Life happens, I definitely understand :-)

**Comment by ryangribble** [comment]
As is the case with preview API's there is also the possibility of shifting goal posts! 

Some changes were recently announced to this that will need to be taken into account too 

https://developer.github.com/changes/2017-03-22-protected-branches-admin-improvements/

@m-zuber did you want to push up the code you do have so far?

**Comment by M-Zuber** [comment]
Based on the tests I tried updating, and the UI, any `IncludeAdmin` is ignored if it is not at the top level of the `BranchProtectionSettings`
Should I just make those changes as part of this PR, or make a new one that updates the existing code in master, and then rebase this PR on top of it?

**Comment by M-Zuber** [comment]
To be more accurate, it seems that (as per docs I suppose) if any of the update objects pass in `true` for `includeAdmins` the top level setting is set as `true`

**Comment by M-Zuber** [comment]
@shiftkey or @ryangribble any tips on how to proceed with this?
The main issue is that we need to obsolete all ctors that accept the `include_admins` param, as setting it in one object can be overriden by another, or by the top level param (which still needs to be added)

**Comment by ryangribble** [comment]
Hope you don't mind @M-Zuber but I've taken over this PR 😀

I realized that [we don't have much time](https://developer.github.com/changes/2017-06-16-loki-preview-ending-soon/) before we have to pass the new `required_pull_request_reviews` field in all branch protection calls, otherwise they will stop working when the API becomes official in less than a week

First post has been updated...

**Comment by ryangribble** [comment]
OK this should be good!  All integration tests are passing.  

Just needs a sanity check for typos etc

**Comment by M-Zuber** [comment]
Thanks for picking this one up, and my apologies for dropping the ball

File: Octokit/Models/Request/BranchProtectionUpdate.cs [hunk 1]
```diff
@@ -304,28 +293,146 @@ internal string DebuggerDisplay
...
+        {
+            if (enabled)
+            {
+                Teams = new BranchProtectionTeamCollection();
+                Users = new BranchProtectionUserCollection();
+            }
+            else
+            {
+                // API requires an empty array to be passed, to disable these checks
+                Teams = null;
```
Comment (M-Zuber) [inline Octokit/Models/Request/BranchProtectionUpdate.cs]: Is there a way to make it clearer what is happening here?
as a newcomer to the codebase might be confused by `null` being equated with an empty array

**Comment by M-Zuber** [comment]
One little thought, but otherwise LGTM
3oEdva9BUHPIs2SkGk

File: Octokit/Models/Request/BranchProtectionUpdate.cs (hunk 1 above)
Comment (ryangribble) [reply Octokit/Models/Request/BranchProtectionUpdate.cs]: Yep I agree, this is not great but it's the best I could come up with 😡 

Im not sure why this API endpoint is different to the other branch protection ones!  Normally to "disable" a protection setting, you can send a `null` for the object but in this case to disable the item, you have to send an empty bodied member.

So basically to be disabled, as per the following screen shot, we have to send an instance of the class but with no members (ie the Teams/Users array are `null`)
image

To enable the dismiss restriction and have only admins (as per next screenshot) you send empty Teams/Users arrays.
image

To specify teams/users you populate them
image



So given that we need to somehow have either the empty bodied class or empty array or populated arrays, what i came up with was the `ctor` thqat takes a `bool` for whether the restriction is enabled (admin only, thus empty team/user list)... or disabled (`null` arrays, so the class has an empty body).

I tried to explain the usage in the xmldoc comments which users hopefully see in intellisense.

If you have any better ideas for implementation let me know!

Comment (M-Zuber) [reply Octokit/Models/Request/BranchProtectionUpdate.cs]: From the perspective of an external user, I think what you have is a nice solution.
My worry is more on someone coming along later trying to contribute, and not understanding the code comment
Maybe something like `API requires an object with empty members to be passed in order to disable these checks`?

Comment (ryangribble) [reply Octokit/Models/Request/BranchProtectionUpdate.cs]: Ah right, sorry if misunderstood. Yes that code comment is poorly worded, I will update thanks!

**Comment by ryangribble** [comment]
release_notes:  Implement `RequiredPullRequestReviews` support in `IRepositoryBranchesClient` BranchProtection methods, required when [preview mode ends on 1st September 2017](https://developer.github.com/changes/2017-06-16-loki-preview-ending-soon/)

**Comment by ryangribble** [comment]
advisories: This release contains the necessary Octokit changes to ensure that we send the `required-pull-request-reviews` field which becomes mandatory  when the Protected Branches API [graduates from preview mode](https://developer.github.com/changes/2017-06-16-loki-preview-ending-soon/) on the 1st September

**Pull request merged**